tests/generators/generators.js

```diff
@@ -0,0 +1,107 @@
+function *stmt_yield(): Generator<number, void, void> {
+  yield 0; // ok
+  yield ""; // error: string ~> number
+}
+
+function *stmt_next(): Generator<void, void, number> {
+  (yield undefined : number); // ok
+  (yield undefined : string); // error: number ~> string
+}
+
+function *stmt_return_ok(): Generator<void, number, void> {
+  return 0; // ok
+}
+
+function *stmt_return_err(): Generator<void, number, void> {
+  return ""; // error: string ~> number
+}
+
+function *infer_stmt() {
+  var x: boolean = yield 0;
+  return "";
+}
+for (var x of infer_stmt()) { (x : string) } // error: number ~> string
+var infer_stmt_next = infer_stmt().next(0).value; // error: number ~> boolean
+if (typeof infer_stmt_next === "undefined") {
+} else if (typeof infer_stmt_next === "number") {
+} else {
+  (infer_stmt_next : boolean) // error: string ~> boolean
+}
+
+declare function *decl(): Generator<number, string, boolean>;
+for (var x of decl()) { (x : string) } // error: number ~> string
+var decl_next = decl().next(0).value; // error: number ~> boolean
+if (typeof decl_next === "undefined") {
+} else if (typeof decl_next === "number") {
+} else {
+  (decl_next : boolean) // error: string ~> boolean
+}
+
+function *widen_next() {
+  var x = yield 0;
+  if (typeof x === "number") {
+  } else if (typeof x === "boolean") {
+  } else {
+    (x : string) // ok, sherlock
+  }
+}
+widen_next().next(0)
+widen_next().next("")
+widen_next().next(true)
+
+function *widen_yield() {
+  yield 0;
+  yield "";
+  yield true;
+}
+for (var x of widen_yield()) {
+  if (typeof x === "number") {
+  } else if (typeof x === "boolean") {
+  } else {
+    (x : string) // ok, sherlock
+  }
+}
+
+function *delegate_next_generator() {
+  function *inner() {
+    var x: number = yield undefined; // error: string ~> number
+  }
+  yield *inner();
+}
+delegate_next_generator().next("");
+
+function *delegate_yield_generator() {
+  function *inner() {
+    yield "";
+  }
+
+  yield *inner();
+}
+for (var x of delegate_yield_generator()) {
+  (x : number) // error: string ~> number
+}
+
+function *delegate_return_generator() {
+  function *inner() {
+    return "";
+  }
+
+  var x: number = yield *inner() // TODO: error: string ~> number
+}
+
+// only generators can make use of a value passed to next
+function *delegate_next_iterable(xs: Array<number>) {
+  yield *xs;
+}
+delegate_next_iterable([]).next(""); // TODO: error: string ~> void
```

Comment:
I see what you mean about `Iterator`'s `next` taking no arguments. I agree that passing too many arguments should be OK.

I need to think on this a bit. In particular, I'm curious about generators like this:

```
function *hmm(xx: Array<number>): Generator<number, bool, *> {
  var str = yield 42;
  yield *xs;
  return str === "hello";
}
```

What is the type of `*` here? Is it `string` or `string | void`? If we choose `string`, then I would need to always call `next` with an argument, when that only matters once.

...continuing to type as I think, I'm realizing that the first call to `next` is always going to have a void argument, and indeed the argument to next is optional already...

OK, I think I'm convinced. Still going to think this over, but right now I think they way forward is to mark this test as OK and try to get the error to disappear.

Separately, thanks for the hint on use types for this. Sounds like a fun challenge.